crates/sui/src/upgrade_compatibility.rs

```diff
@@ -378,36 +381,159 @@ pub(crate) async fn check_compatibility(
+    let mut writer;
+    if enable_colors {
+        writer = codespan_reporting::term::termcolor::Buffer::ansi();
+    } else {
+        writer = codespan_reporting::term::termcolor::Buffer::no_color();
+    }
+    let mut file_id_map = HashMap::new();
+
+    for (name, err) in errors {
+        let compiled_unit_with_source = new_package
+            .package
+            .get_module_by_name_from_root(&name.to_string())
+            .context("Unable to get module")?;
+
+        let source_path = compiled_unit_with_source.source_path.to_string_lossy();
+        let file_id = match file_id_map.get(&source_path) {
+            Some(file_id) => *file_id,
             None => {
-                Err(anyhow!("Module {} is missing from the package", name))?;
+                let source = fs::read_to_string(&compiled_unit_with_source.source_path)
+                    .context("Unable to read source file")?;
+                let file_id = files.add(source_path.clone(), source);
+                file_id_map.insert(source_path.clone(), file_id);
+                file_id
             }
+        };
+
+        codespan_reporting::term::emit(
+            &mut writer,
+            &config,
+            &files,
+            &diag_from_error(err, compiled_unit_with_source, file_id),
+        )?;
+    }
+
+    Err(anyhow!(
+        "Upgrade compatibility check failed:\n{}",
+        String::from_utf8(writer.into_inner()).context("Unable to convert buffer to string")?
+    ))
+}
+
+/// Convert an error to a diagnostic using the specific error type's function.
+fn diag_from_error(
+    error: UpgradeCompatibilityModeError,
+    compiled_unit_with_source: &CompiledUnitWithSource,
+    file_id: usize,
+) -> Diagnostic<usize> {
+    match error {
+        UpgradeCompatibilityModeError::StructMissing { name, .. } => missing_definition_diag(
+            "struct",
+            name.to_string(),
+            compiled_unit_with_source,
+            file_id,
+        ),
+        UpgradeCompatibilityModeError::EnumMissing { name, .. } => {
+            missing_definition_diag("enum", name.to_string(), compiled_unit_with_source, file_id)
         }
+        UpgradeCompatibilityModeError::FunctionMissingPublic { name, .. } => {
+            missing_definition_diag(
+                "public function",
+                name.to_string(),
+                compiled_unit_with_source,
+                file_id,
+            )
+        }
+        UpgradeCompatibilityModeError::FunctionMissingEntry { name, .. } => {
+            missing_definition_diag(
+                "entry function",
+                name.to_string(),
+                compiled_unit_with_source,
+                file_id,
+            )
+        }
+        _ => todo!("Implement diag_from_error for {:?}", error),
     }
+}
 
-    Ok(())
+/// Return a diagnostic for a missing definition.
+fn missing_definition_diag(
+    declaration_kind: &str,
+    identifier_name: String,
```

Comment:
In this case, this could just be `&Identifier` -- that type implements `Display`, so you can pass it into the format message directly, and avoid the copy.